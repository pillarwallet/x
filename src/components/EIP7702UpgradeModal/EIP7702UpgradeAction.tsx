import { isNaN } from 'lodash';
import { TailSpin } from 'react-loader-spinner';

// images
import logoEthereum from '../../assets/images/logo-ethereum.png';
import UpgradeGreenLogo from '../../assets/images/upgrade-green-logo.svg';

// utils
import { limitDigitsNumber } from '../../utils/number';

interface EIP7702UpgradeActionProps {
  onClose: () => void;
  handleUpgrade: () => void;
  gasFeeEstimates?: string;
  isCheckingGas?: boolean;
  hasEnoughGas?: boolean;
}

const CheckIcon = () => {
  return (
    <div className="flex items-center justify-center w-4 h-4 rounded-full bg-[#8A77FF] border-[1.5px] border-[#A594FF]">
      <svg
        className="w-3 h-3 text-white"
        fill="currentColor"
        viewBox="0 0 20 20"
        strokeWidth="2"
      >
        <path
          fillRule="evenodd"
          d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
          clipRule="evenodd"
        />
      </svg>
    </div>
  );
};

const EIP7702UpgradeAction = ({
  onClose,
  gasFeeEstimates,
  handleUpgrade,
  isCheckingGas = false,
  hasEnoughGas = false,
}: EIP7702UpgradeActionProps) => {
  return (
    <>
      <div className="flex flex-col gap-2.5 items-center">
        <img
          src={UpgradeGreenLogo}
          alt="upgrade-green-arrows"
          className="w-[66px] h-[44px]"
        />
        <p className="text-white font-normal text-xl">Upgrade your account</p>
        <div className="flex justify-center items-center text-sm font-normal gap-1">
          <p className="text-white">Network:</p>
          <img
            src={logoEthereum}
            className="rounded-full w-4 h-4 flex-shrink-0"
            alt="logo-ethereum"
          />
          <p className="text-[#627EEA]">Ethereum</p>
        </div>
      </div>
      <div className="flex flex-col gap-2">
        <div className="flex items-center gap-[7px]">
          <CheckIcon />
          <p className="text-white text-base font-normal">
            <span className="text-white" style={{ opacity: 0.7 }}>
              Faster Transactions.{' '}
            </span>
            Lower fees.
          </p>
        </div>
        <div className="flex items-center gap-[7px]">
          <CheckIcon />
          <p className="text-white text-base font-normal">
            <span className="text-white" style={{ opacity: 0.7 }}>
              Pay with any token.{' '}
            </span>
            Any time.
          </p>
        </div>
        <div className="flex items-center gap-[7px]">
          <CheckIcon />
          <p className="text-white text-base font-normal">
            <span className="text-white" style={{ opacity: 0.7 }}>
              Same account.{' '}
            </span>
            Smart Features.
          </p>
        </div>
      </div>
      <div className="flex flex-col gap-2 w-full items-center">
        <div className="flex gap-2 items-center">
          <p
            className="text-[13px] text-white font-normal"
            style={{ opacity: 0.7 }}
          >
            Fees:
          </p>
          {isCheckingGas ? (
            <TailSpin color="#FFFFFF" height={12} width={12} />
          ) : (
            <p className="text-[13px] text-white font-normal">
              {gasFeeEstimates && !isNaN(Number(gasFeeEstimates))
                ? limitDigitsNumber(Number(gasFeeEstimates))
                : '-'}{' '}
              ETH
            </p>
          )}
        </div>
        <div className="w-full rounded-[10px] bg-[#121116] p-[2px_2px_6px_2px]">
          <button
            className={`flex items-center justify-center w-full rounded-[8px] h-[42px] p-[1px_6px_1px_6px] text-white font-normal text-base transition-opacity ${
              isCheckingGas || !hasEnoughGas
                ? 'bg-[#3D3A44] cursor-not-allowed opacity-60'
                : 'bg-[#8A77FF] hover:opacity-90'
            }`}
            type="button"
            onClick={handleUpgrade}
            disabled={isCheckingGas || !hasEnoughGas}
          >
            {!hasEnoughGas && !isCheckingGas ? 'Not enough gas' : 'Upgrade now'}
          </button>
        </div>
        <button
          type="button"
          className="text-white text-xs font-normal underline underline-offset-2 opacity-70"
          onClick={onClose}
        >
          Upgrade later
        </button>
      </div>
    </>
  );
};

export default EIP7702UpgradeAction;
