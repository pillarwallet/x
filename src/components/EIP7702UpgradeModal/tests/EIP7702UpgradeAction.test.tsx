/* eslint-disable react/jsx-props-no-spreading */
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// components
import EIP7702UpgradeAction from '../EIP7702UpgradeAction';

// Mock dependencies
vi.mock('../../utils/number', () => ({
  limitDigitsNumber: vi.fn((num: number) => num),
}));

vi.mock('react-loader-spinner', () => ({
  TailSpin: () => <div data-testid="tail-spin">Loading</div>,
}));

vi.mock('../../assets/images/logo-ethereum.png', () => ({
  default: 'logo-ethereum.png',
}));

vi.mock('../../assets/images/upgrade-green-logo.svg', () => ({
  default: 'upgrade-green-logo.svg',
}));

const baseProps = {
  onClose: vi.fn(),
  handleUpgrade: vi.fn(),
  gasFeeEstimates: '0.001',
  isCheckingGas: false,
  hasEnoughGas: true,
};

describe('<EIP7702UpgradeAction />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('Rendering', () => {
    it('renders correctly and matches snapshot', () => {
      const tree = renderer
        .create(<EIP7702UpgradeAction {...baseProps} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('displays main elements correctly', () => {
      render(<EIP7702UpgradeAction {...baseProps} />);

      expect(screen.getByText('Upgrade your account')).toBeInTheDocument();
      expect(screen.getByText('Network:')).toBeInTheDocument();
      expect(screen.getByText('Ethereum')).toBeInTheDocument();
      expect(screen.getByText('Faster Transactions.')).toBeInTheDocument();
      expect(screen.getByText('Lower fees.')).toBeInTheDocument();
      expect(screen.getByText('Pay with any token.')).toBeInTheDocument();
      expect(screen.getByText('Any time.')).toBeInTheDocument();
      expect(screen.getByText('Same account.')).toBeInTheDocument();
      expect(screen.getByText('Smart Features.')).toBeInTheDocument();
      expect(screen.getByText('Upgrade now')).toBeInTheDocument();
      expect(screen.getByText('Upgrade later')).toBeInTheDocument();
    });

    it('displays gas fee when provided', () => {
      render(<EIP7702UpgradeAction {...baseProps} gasFeeEstimates="0.001" />);

      expect(screen.getByText('Fees:')).toBeInTheDocument();
      expect(screen.getByText(/0\.001/)).toBeInTheDocument();
      expect(screen.getByText(/ETH/)).toBeInTheDocument();
    });

    it('shows loading spinner when checking gas', () => {
      render(<EIP7702UpgradeAction {...baseProps} isCheckingGas />);

      expect(screen.getByTestId('tail-spin')).toBeInTheDocument();
      expect(screen.queryByText('ETH')).not.toBeInTheDocument();
    });
  });

  describe('Benefits display', () => {
    it('displays all three benefits with checkmarks', () => {
      const { container } = render(<EIP7702UpgradeAction {...baseProps} />);

      // Check icons should be present
      const checkIcons = container.querySelectorAll('[class*="bg-"]');
      expect(checkIcons.length).toBeGreaterThanOrEqual(3);

      // All benefit text should be visible
      expect(screen.getByText('Faster Transactions.')).toBeInTheDocument();
      expect(screen.getByText('Lower fees.')).toBeInTheDocument();
      expect(screen.getByText('Pay with any token.')).toBeInTheDocument();
      expect(screen.getByText('Any time.')).toBeInTheDocument();
      expect(screen.getByText('Same account.')).toBeInTheDocument();
      expect(screen.getByText('Smart Features.')).toBeInTheDocument();
    });
  });

  describe('Button states', () => {
    it('shows "Upgrade now" when has enough gas', () => {
      render(<EIP7702UpgradeAction {...baseProps} hasEnoughGas />);

      const button = screen.getByText('Upgrade now');
      expect(button).toBeInTheDocument();
      expect(button.closest('button')).not.toBeDisabled();
    });

    it('shows "Not enough gas" when has insufficient gas', () => {
      render(<EIP7702UpgradeAction {...baseProps} hasEnoughGas={false} />);

      const button = screen.getByText('Not enough gas');
      expect(button).toBeInTheDocument();
      expect(button.closest('button')).toBeDisabled();
    });

    it('disables button when checking gas', () => {
      render(<EIP7702UpgradeAction {...baseProps} isCheckingGas />);

      const button = screen.getByText('Upgrade now');
      expect(button.closest('button')).toBeDisabled();
    });

    it('disables button when not enough gas', () => {
      render(<EIP7702UpgradeAction {...baseProps} hasEnoughGas={false} />);

      const button = screen.getByText('Not enough gas');
      expect(button.closest('button')).toBeDisabled();
    });
  });

  describe('User interactions', () => {
    it('calls handleUpgrade when upgrade button is clicked', () => {
      render(<EIP7702UpgradeAction {...baseProps} />);

      const button = screen.getByText('Upgrade now');
      fireEvent.click(button);

      expect(baseProps.handleUpgrade).toHaveBeenCalledTimes(1);
    });

    it('calls onClose when "Upgrade later" is clicked', () => {
      render(<EIP7702UpgradeAction {...baseProps} />);

      const laterButton = screen.getByText('Upgrade later');
      fireEvent.click(laterButton);

      expect(baseProps.onClose).toHaveBeenCalledTimes(1);
    });

    it('does not call handleUpgrade when button is disabled', () => {
      render(<EIP7702UpgradeAction {...baseProps} hasEnoughGas={false} />);

      const button = screen.getByText('Not enough gas');
      fireEvent.click(button);

      expect(baseProps.handleUpgrade).not.toHaveBeenCalled();
    });
  });

  describe('Edge cases', () => {
    it('handles missing gas fee estimates', () => {
      render(
        <EIP7702UpgradeAction {...baseProps} gasFeeEstimates={undefined} />
      );

      expect(screen.getByText('Fees:')).toBeInTheDocument();
      expect(
        screen.getByText((content) => content.includes('-'))
      ).toBeInTheDocument();
    });

    it('handles invalid gas fee estimates', () => {
      render(<EIP7702UpgradeAction {...baseProps} gasFeeEstimates="invalid" />);

      expect(screen.getByText('Fees:')).toBeInTheDocument();
      expect(
        screen.getByText((content) => content.includes('-'))
      ).toBeInTheDocument();
    });

    it('handles empty gas fee estimates', () => {
      render(<EIP7702UpgradeAction {...baseProps} gasFeeEstimates="" />);

      expect(screen.getByText('Fees:')).toBeInTheDocument();
      expect(
        screen.getByText((content) => content.includes('-'))
      ).toBeInTheDocument();
    });
  });

  describe('Default props', () => {
    it('uses default values for optional props', () => {
      const minimalProps = {
        onClose: vi.fn(),
        handleUpgrade: vi.fn(),
      };

      render(<EIP7702UpgradeAction {...minimalProps} />);

      // Should render without errors
      expect(screen.getByText('Upgrade your account')).toBeInTheDocument();
      expect(screen.getByText('Upgrade later')).toBeInTheDocument();
    });
  });

  describe('Button styling', () => {
    it('applies active button styles when enabled', () => {
      const { container } = render(<EIP7702UpgradeAction {...baseProps} />);

      const button = container.querySelector('.bg-\\[\\#8A77FF\\]');
      expect(button).toBeInTheDocument();
    });

    it('applies disabled button styles when disabled', () => {
      const { container } = render(
        <EIP7702UpgradeAction {...baseProps} hasEnoughGas={false} />
      );

      const button = container.querySelector('.bg-\\[\\#3D3A44\\]');
      expect(button).toBeInTheDocument();
    });
  });
});
