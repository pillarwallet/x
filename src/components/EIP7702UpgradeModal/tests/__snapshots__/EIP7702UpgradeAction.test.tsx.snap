// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EIP7702UpgradeAction /> > Rendering > renders correctly and matches snapshot 1`] = `
[
  <div
    className="flex flex-col gap-2.5 items-center"
  >
    <img
      alt="upgrade-green-arrows"
      className="w-[66px] h-[44px]"
      src="data:image/svg+xml,%3csvg%20width='66'%20height='44'%20viewBox='0%200%2066%2044'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M27.5%2019.25V41.6429C27.5%2042.9447%2028.5553%2044%2029.8571%2044H36.1429C37.4447%2044%2038.5%2042.9447%2038.5%2041.6429V19.25H44.3751C46.3889%2019.25%2047.4753%2016.8879%2046.1647%2015.3589L34.7897%202.08796C33.849%200.990443%2032.151%200.990442%2031.2103%202.08795L19.8353%2015.3588C18.5247%2016.8879%2019.6111%2019.25%2021.6249%2019.25H27.5Z'%20fill='%235CFF93'/%3e%3cpath%20opacity='0.5'%20d='M7.33333%2028.5312V36.1429C7.33333%2037.4447%208.38866%2038.5%209.69048%2038.5H12.3095C13.6113%2038.5%2014.6667%2037.4447%2014.6667%2036.1429V28.5312H16.6511C18.7006%2028.5312%2019.7737%2026.0962%2018.3907%2024.5836L12.7396%2018.4027C11.8051%2017.3806%2010.1949%2017.3806%209.26036%2018.4027L3.6093%2024.5836C2.22635%2026.0962%203.29943%2028.5312%205.34894%2028.5312H7.33333Z'%20fill='%235CFF93'/%3e%3cpath%20opacity='0.5'%20d='M51.3333%2028.5312V36.1429C51.3333%2037.4447%2052.3887%2038.5%2053.6905%2038.5H56.3095C57.6113%2038.5%2058.6667%2037.4447%2058.6667%2036.1429V28.5312H60.6511C62.7006%2028.5312%2063.7737%2026.0962%2062.3907%2024.5836L56.7396%2018.4027C55.8051%2017.3806%2054.1949%2017.3806%2053.2604%2018.4027L47.6093%2024.5836C46.2264%2026.0962%2047.2994%2028.5312%2049.3489%2028.5312H51.3333Z'%20fill='%235CFF93'/%3e%3c/svg%3e"
    />
    <p
      className="text-white font-normal text-xl"
    >
      Upgrade your account
    </p>
    <div
      className="flex justify-center items-center text-sm font-normal gap-1"
    >
      <p
        className="text-white"
      >
        Network:
      </p>
      <img
        alt="logo-ethereum"
        className="rounded-full w-4 h-4 flex-shrink-0"
        src="/src/assets/images/logo-ethereum.png"
      />
      <p
        className="text-[#627EEA]"
      >
        Ethereum
      </p>
    </div>
  </div>,
  <div
    className="flex flex-col gap-2"
  >
    <div
      className="flex items-center gap-[7px]"
    >
      <div
        className="flex items-center justify-center w-4 h-4 rounded-full bg-[#8A77FF] border-[1.5px] border-[#A594FF]"
      >
        <svg
          className="w-3 h-3 text-white"
          fill="currentColor"
          strokeWidth="2"
          viewBox="0 0 20 20"
        >
          <path
            clipRule="evenodd"
            d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
            fillRule="evenodd"
          />
        </svg>
      </div>
      <p
        className="text-white text-base font-normal"
      >
        <span
          className="text-white"
          style={
            {
              "opacity": 0.7,
            }
          }
        >
          Faster Transactions.
           
        </span>
        Lower fees.
      </p>
    </div>
    <div
      className="flex items-center gap-[7px]"
    >
      <div
        className="flex items-center justify-center w-4 h-4 rounded-full bg-[#8A77FF] border-[1.5px] border-[#A594FF]"
      >
        <svg
          className="w-3 h-3 text-white"
          fill="currentColor"
          strokeWidth="2"
          viewBox="0 0 20 20"
        >
          <path
            clipRule="evenodd"
            d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
            fillRule="evenodd"
          />
        </svg>
      </div>
      <p
        className="text-white text-base font-normal"
      >
        <span
          className="text-white"
          style={
            {
              "opacity": 0.7,
            }
          }
        >
          Pay with any token.
           
        </span>
        Any time.
      </p>
    </div>
    <div
      className="flex items-center gap-[7px]"
    >
      <div
        className="flex items-center justify-center w-4 h-4 rounded-full bg-[#8A77FF] border-[1.5px] border-[#A594FF]"
      >
        <svg
          className="w-3 h-3 text-white"
          fill="currentColor"
          strokeWidth="2"
          viewBox="0 0 20 20"
        >
          <path
            clipRule="evenodd"
            d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
            fillRule="evenodd"
          />
        </svg>
      </div>
      <p
        className="text-white text-base font-normal"
      >
        <span
          className="text-white"
          style={
            {
              "opacity": 0.7,
            }
          }
        >
          Same account.
           
        </span>
        Smart Features.
      </p>
    </div>
  </div>,
  <div
    className="flex flex-col gap-2 w-full items-center"
  >
    <div
      className="flex gap-2 items-center"
    >
      <p
        className="text-[13px] text-white font-normal"
        style={
          {
            "opacity": 0.7,
          }
        }
      >
        Fees:
      </p>
      <p
        className="text-[13px] text-white font-normal"
      >
        0.001
         
        ETH
      </p>
    </div>
    <div
      className="w-full rounded-[10px] bg-[#121116] p-[2px_2px_6px_2px]"
    >
      <button
        className="flex items-center justify-center w-full rounded-[8px] h-[42px] p-[1px_6px_1px_6px] text-white font-normal text-base transition-opacity bg-[#8A77FF] hover:opacity-90"
        disabled={false}
        onClick={[MockFunction spy]}
        type="button"
      >
        Upgrade now
      </button>
    </div>
    <button
      className="text-white text-xs font-normal underline underline-offset-2 opacity-70"
      onClick={[MockFunction spy]}
      type="button"
    >
      Upgrade later
    </button>
  </div>,
]
`;
